text/0006-wasm-smart-contracts.md

```diff
@@ -0,0 +1,440 @@
+---
+layout: default
+title: Wasm_Smart_Contracts
+parent: RFCs
+---
+
+- Feature Name: Support writing Smart Contracts to be executed within a WebAssembly runtime
+- Start Date: 2020-05-20
+- RFC PR: (leave this empty)
+- Fabric Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC is providing:
+
+* a WebAssembly runtime that can be used as a Chaincode implementation in Fabric
+* implementation of a set of APIs to permit Smart Contracts to be written in Rust and then compiled to Wasm bytecode
+* documented interface to permit the community to implement Smart Contracts in other languages. 
+
+A Rust developer would write their Smart Contract with Rust and produce a library. This they would then target to be compiled to Wasm bytecode - the resulting wasm binary would then be deployed using Fabric v2 "External Builders" into a Wasm runtime. This contract is then treated just the same as any other contract or chaincode. 
+
+The APIs will be implemented at a level consistent with the existing Contract APIs but taking the opportunity to make improvements. 
+
+Note: code examples are provided for illustration, and should be considered pseudo-code to outline the functionality that will be produced.
+
+Prototypes are available for the [Rust API](https://github.com/hyperledgendary/fabric-contract-api-rust), [Wasm Chaincode](https://github.com/hyperledgendary/fabric-chaincode-wasm); work is still taking place in these repos so please get contact with questions.
+
+# Motivation
```

Comment:
I would like the understand the problem being solved by this RFC. Currently, we can write smart-contract using go, java, nodejs, evm, and contract-api. Why do we need yet another way to write a smart-contract? Sometimes having too many choices is bad too. Especially when we need to add a new chaincode API, we need to ensure that all repos are updated, works correctly, and being maintained too. 

For example, if it helps in (a) improving the performance compared to go or other supported languages or (b) enabling deterministic execution or (c) enabling new use-cases or (d) avoiding bugs, this RFC can be understood better.

Reply:
We actually want to deprecate the current chaincode model over time.  While the current chaincode implementation made sense at the time given we did not want to write our own virtual machine / byte code interpreter as well as the fact that we did not feel EVM byte code would meet all of our needs, if something like WebAssembly had existed at the time, it's the route we would have taken.

Benefits:
- reduce duplication of code - we no longer have to write shims in multiple languages
- identical model can be used for out-of-process and in-process execution (for out-of-process we'd only need one external host process / shim as well)
- reduce overall complexity of the system

Additionally, the current chaincode stub "interface" has become polluted with functions as things have evolved.  To overcome that complexity, the contract API wrapper layer was created which while making things more usable, actually makes the code base more complex.  So we are not simply going to "port" the existing stub / contract API to this implementation but rather create a new set of APIs which better map to the overall Fabric model.

Reply:
Thanks, Gari. I wasn't aware of the whole plan. Your response answered both of my comments and now I understand the goal of this RFC.

Reply:
great comments / questions ... needed to be answered for sure